# Отчёт_08_02_2021

## Изменение главной директории (выполнять только один раз!)

In [1]:
import sys
import os
os.chdir('..//')
print(os.getcwd())

C:\Users\NightHegg\Desktop\projects\SRW


## Подключение необходимых библиотек

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import scipy as sp

from scr.class_basic_method import basic_method
from scr.class_schwarz_multiplicative import schwarz_multiplicative
from scr.class_schwarz_additive import schwarz_additive
from scr.class_schwarz_two_level_additive import schwarz_two_level_additive

## Функция для получения результатов

In [3]:
def calculate_iters(cur_task, cur_method, list_mesh, list_subds):
    dict_result = {}
    dict_result_sigma = {}
    for cur_mesh in list_mesh:
        dict_iters_temp = {}
        dict_sigma_temp = {}
        for cur_amnt_subds in list_subds:
            obj = cur_method(cur_task, cur_mesh, cur_amnt_subds, coef_convergence)
            obj.get_solution()
            
            array_amnt_subds = np.prod(np.array(cur_amnt_subds))
            styled_amnt_subds = f"{array_amnt_subds} области" if array_amnt_subds < 5 else f"{array_amnt_subds} областей"
            
            dict_iters_temp[styled_amnt_subds] = obj.amnt_iterations
            dict_sigma_temp[styled_amnt_subds] = obj.get_special_sigma()
            
        dict_result[f"{cur_mesh**2} точек"] = dict_iters_temp
        dict_result_sigma[f"{cur_mesh**2} точек"] = dict_sigma_temp
    return dict_result, dict_result_sigma


def calculate_conv(cur_task, cur_method, cur_mesh, cur_subds, list_coef):
    dict_result = {}
    dict_result_temp = {}
    for cur_coef in list_coef:
        obj = cur_method(cur_task, cur_mesh, cur_subds, cur_coef)
        obj.get_solution()
        dict_result_temp[f"{cur_coef:.0e}"] = obj.amnt_iterations
    dict_result["Количество итераций"] = dict_result_temp
    return dict_result

### Предоставленные задачи:

- закрепление с трёх сторон;
- закрепление с двух сторон;
<div>
<img src="img/tasks_Rodin.jpg" width="250"/>
</div>

## Первая задача:

In [4]:
cur_task = 1
coef_convergence = 1e-5

list_subds = [[2, 1], [4, 1], [8, 1]]
list_mesh = [20, 25, 30]

test_mesh = 25
test_subds = [4, 1]
list_coef = [1e-3, 1e-4, 1e-5]

### Мультипликативный метод Шварца:

In [5]:
dict_iters_sm, dict_sigma_sm = calculate_iters(cur_task, schwarz_multiplicative, list_mesh, list_subds)

Таблица количества итераций:

In [6]:
pd.DataFrame.from_dict(dict_iters_sm)

400 точек  625 точек  900 точек
2 области          12         13         12
4 области          31         31         32
8 областей        114         92        111

Таблица разницы для осевого напряжения:

In [7]:
pd.DataFrame.from_dict(dict_sigma_sm)

400 точек                       625 точек  \
2 области   min = 3.14e-06, max = 6.75e-07  min = 1.68e-06, max = 4.71e-07   
4 области   min = 6.77e-06, max = 1.95e-06  min = 5.87e-06, max = 1.15e-06   
8 областей  min = 2.67e-05, max = 1.32e-05  min = 1.61e-05, max = 4.28e-06   

                                 900 точек  
2 области   min = 8.85e-07, max = 1.79e-07  
4 области   min = 6.83e-06, max = 1.97e-06  
8 областей  min = 3.32e-05, max = 1.33e-05

#### Дополнительные задачи:

In [8]:
dict_conv_sm = calculate_conv(cur_task, schwarz_multiplicative, test_mesh, test_subds, list_coef)

Таблица зависимости количества итераций от критерия сходимости (625 точек, 4 подобласти):

In [9]:
pd.DataFrame(dict_conv_sm)

Количество итераций
1e-03                   15
1e-04                   23
1e-05                   31

### Аддитивный метод Шварца:

In [10]:
dict_iters_sa, dict_sigma_sa = calculate_iters(cur_task, schwarz_additive, list_mesh, list_subds)

Таблица количества итераций:

In [11]:
pd.DataFrame.from_dict(dict_iters_sa)

400 точек  625 точек  900 точек
2 области          22         23         21
4 области          40         39         44
8 областей        145        102        143

Таблица разницы для осевого напряжения:

In [12]:
pd.DataFrame.from_dict(dict_sigma_sa)

400 точек                       625 точек  \
2 области   min = 2.93e-05, max = 3.65e-05  min = 2.39e-06, max = 2.14e-05   
4 области   min = 2.26e-05, max = 3.41e-05  min = 9.50e-06, max = 2.43e-05   
8 областей  min = 6.11e-05, max = 1.66e-04  min = 3.73e-05, max = 8.02e-05   

                                 900 точек  
2 области   min = 2.47e-06, max = 1.29e-05  
4 области   min = 2.59e-05, max = 3.64e-05  
8 областей  min = 4.88e-05, max = 1.60e-04

#### Дополнительные задачи:

In [13]:
dict_conv_sa = calculate_conv(cur_task, schwarz_additive, test_mesh, test_subds, list_coef)

Таблица зависимости количества итераций от критерия сходимости (625 точек, 4 подобласти):

In [14]:
pd.DataFrame(dict_conv_sa)

Количество итераций
1e-03                   20
1e-04                   29
1e-05                   39

### Двухуровневый аддитивный метод Шварца:

In [15]:
dict_iters_tsa, dict_sigma_tsa = calculate_iters(cur_task, schwarz_two_level_additive, list_mesh, list_subds)

Таблица количества итераций:

In [16]:
pd.DataFrame.from_dict(dict_iters_tsa)

400 точек  625 точек  900 точек
2 области          13         13         13
4 области          15         14         15
8 областей         27         23         26

Таблица разницы для осевого напряжения:

In [17]:
pd.DataFrame.from_dict(dict_sigma_tsa)

400 точек                       625 точек  \
2 области   min = 2.22e-05, max = 2.57e-05  min = 2.78e-05, max = 3.42e-05   
4 области   min = 3.41e-05, max = 3.78e-05  min = 5.48e-05, max = 1.10e-04   
8 областей  min = 4.33e-04, max = 4.38e-04  min = 5.25e-05, max = 6.61e-05   

                                 900 точек  
2 области   min = 2.35e-05, max = 1.93e-05  
4 области   min = 6.07e-05, max = 8.04e-05  
8 областей  min = 2.23e-04, max = 4.39e-04

#### Дополнительные задачи:

In [18]:
dict_conv_tsa = calculate_conv(cur_task, schwarz_two_level_additive, test_mesh, test_subds, list_coef)

Таблица зависимости количества итераций от критерия сходимости (625 точек, 4 подобласти):

In [19]:
pd.DataFrame(dict_conv_tsa)

Количество итераций
1e-03                    7
1e-04                   11
1e-05                   14